Improve protocol diagnostics

nilsvu opened the pull request:
## Proposed changes

Closes #2205.

Require that protocols implement both a SFINAE-friendly unary type trait
and a diagnostic metafunction that triggers useful compiler errors
to understand why a type does not conform to the protocol. With this
change, protocol conformance failures should always trigger errors messages
that are as good or better as the standard compiler erros.

### Types of changes:

- [ ] Bugfix
- [x] New feature
- [ ] Refactor

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsdeppe reviewed (commented):
Please just remove the numerical flux protocol. I've spent over an hour now fighting with the bad diagnostics. I view protocols as a failed experiment that we need to remove ASAP. If you want to move it into the elliptic code that's fine by me, but I want it out of the evolution code.

nilsvu committed:
Improve protocol diagnostics

Require that protocols implement a diagnostic metafunction that triggers
useful compiler errors to understand why a type does not conform to the
protocol. With this change, protocol conformance failures should always
trigger errors messages that are as good or better as the standard
compiler erros.

nilsvu commented:
Per our discussion in the Monday meeting a while back, here are some examples of how the error messages look like with this PR (see below). I'd like to stress that with this PR the error messages will be as good or better than the standard compiler errors, since the protocol's testing metafunction just emits the standard compiler errors plus any additional static asserts the user may have added. And I think anything that improves our compiler errors can be quite useful in our code, given its tendency to dump massive compiler backtraces on the user 😄 These are some of the advantages I see with protocols (**edit:** with the improvements made in this PR):
- The compile-time errors can be caught earlier in static asserts for protocol conformance
- They can have more useful error messages that we have control over, or else fall back to the standard compiler errors
- The compile-time requirements are documented in the protocol documentation, including an example from the protocol test

Here are the examples:

#### Example 1: A numerical flux has no `argument_tags`

```
error: no type named 'argument_tags' in 'ScalarWave::PenaltyFlux<1>'
    using argument_tags = typename ConformingType::argument_tags;
```

#### Example 2: The call operator of the numerical flux has an argument of the wrong type

```
error: no matching function for call to object of type 'ScalarWave::PenaltyFlux<1>'
  using type = decltype(std::declval<NumericalFluxType>()(
...
note: candidate function not viable: no known conversion from 'not_null<Tensor<DataVector, brigand::list<>, brigand::list<> > *>' to 'not_null<int *>' for 1st argument
  void operator()(
```

#### Example 3: The call operator doesn't return `void` for some reason

```
error: static_assert failed due to requirement 'std::is_same_v<int, void>' "The 'operator()' must return 'void'."
    static_assert(
```

#### Example 4: A package tag is missing the `type`

(In the first commit I remove the requirement that the package tags are DataBox tags per #2204)

```
error: no type named 'type' in 'ScalarWave::PenaltyFlux<1>::NormalTimesVPlus'
  template<typename T> using type_from = typename T::type;
```

---

All of these examples also emit a line like this:

```
note: in instantiation of variable template specialization 'tt::assert_conforms_to<ScalarWave::PenaltyFlux<1>, dg::protocols::NumericalFlux, true>' requested here
    tt::assert_conforms_to<ConformingType, Protocol, true>;
```

This should give the indication that the class `ScalarWave::PenaltyFlux<1>` fails to conform to `dg::protocols::NumericalFlux` and the user can look up the protocol documentation for details if the compiler error is not enough to understand what's going on.

I think this PR addresses the issue with insufficient (non-existent) compiler errors when protocol conformance checks fail #2205. Please let me know if anything is unclear!

nilsdeppe commented:
These errors are a lot clearer and I'm willing to give this a try. I don't really agree with your arguments for protocols.

- Catching errors earlier was actually what made it impossible for me to debug the issue. The static assert did not tell me which of the 200 lines of code was wrong, just that something was wrong. It does avoid having compilers hog all the memory on the system, but that I think isn't exactly a great trade off for not getting a useful error message. I realize that this PR is improving the errors, but please don't pretend that protocols inherently offers such an improvement. It's up to the author of the protocol to make the protocol better than the default compiler errors.
- They do **not** fall back to the compiler error message. My entire issue with what's in develop is that they blocked what would've been the compiler error messages that told me what was wrong. I think this is actually a downside of protocols. (I do not know of a way to remedy this since it is sort of the point)
- I do think having a systematic way of documenting interface requirements, such as what protocols offers, is a good thing.

nilsdeppe commented on src/NumericalAlgorithms/DiscontinuousGalerkin/Protocols.hpp:
```diff
@@ -107,16 +123,32 @@ struct IsNumericalFluxCallable
+      detail::IsNumericalFluxCallable<ConformingType>, std::false_type>;
+
+  template <typename ConformingType>
+  struct test : std::true_type {
```
Why does this inherit off of `std::true_type`?

nilsdeppe reviewed (changes requested):
Looking through this, I thought protocols was originally merged under the condition that switching to concepts would be easy. This was achieved by having the protocols be unary type traits. It appears this is no longer true, which means converting to concepts would be a lot more work. I don't currently see what the plan is for migrating to concepts. Is there one? If not, I think we should just remove protocols and wait for concepts, or I need to be convinced that protocols will provide significant value even once concepts is available. That is, we will want to continue using protocols even after concepts is available.

Another concern: it's now clear that at least *sometimes* you get compiler errors that were just as bad as the ones you get previously (yay?), which sort of addresses my complaint that I had no idea what was wrong. What's not clear to me is how do I trigger those instead of the assert "something is wrong"? And should the "something is wrong" provide instructions on how to trigger the more detailed errors? Is there a protocol-agnostic way to trigger the detailed asserts? Do I need to build the tests to trigger the detailed asserts? What if I can't build the test that provides the nice errors because the dependency order requires I build something before it that blocks the individual errors and slams me with the `something is bad`?

Since the error messages now fall back to basically the default compiler errors, how does this not go against one of the primary design goals of protocols? That is, the error messages you shared are great (for me at least), but aren't they still buried in thousands of lines like the actual error would be? If so, what is the advantage protocols offers? One advantage would be that Clang is less likely to spin forever trying to get things to build and instead hits either a static assert or one of the detailed errors, but I don't know if that's true.

nilsvu commented:
In response to your comment I thought more about the design and intention of protocols and significantly stripped down the implementation. So I'll reply in the context of the new fixup commit.

I originally built protocols with the intention of giving our compile-time interfaces structure and documentation. The idea is that a protocol defines a set of requirements (just like a C++20 concept) and a class can indicate that it fulfills these requirements. This is different to concepts where the class that fulfills the requirements doesn't necessarily know it satisfies a particular concept. Talking to you we came to the conclusion that protocols can 1) help us move towards concepts and constraints while we're still at C++17 and 2) complement concepts and constraints once we're at C++20. E.g. any numerical flux class indicates it _is_ a numerical flux, not just silently fulfills the requirements for the concept. Once we're transitioning to concepts and constraints, one option is to have a `tt::ConformsTo` concept and use it in template parameter constraints.

I think the current implementation of protocols on `develop` tries too hard to emulate constraints without proper language support. In this PR (latest fixup) I remove the part of the protocols implementation that tries to check the requirements in a SFINAE-friendly way, because without language support for constraints those checks are opaque to the compiler and lead to the "something is wrong" error that you mention. Instead, I strip down the implementation to the original idea: any class that derives off `tt::ConformsTo<TheProtocol>` is assumed to conform to the protocol and the `tt::conforms_to` unary type trait is provided to check for protocol conformance in a SFINAE context (so SFINAE conformance checks are still supported!). In unit tests and where SFINAE is not needed the `tt::assert_conforms_to` metafunction can be used that checks the requirements in as much detail as the protocol author implemented them, e.g. using static asserts.

In this stripped-down version protocols are a structured way of implementing functionality like the `MarkAsAnalyticSolution` classes we have in the code, plus the ability to perform compile-time checks on classes that are marked as such. Here's some of the advantages I hope we gain with protocols:
- By having `tt::assert_conforms_to` checks as early as possible in compilation units I hope we can reduce compiler backtraces and also better point the user to the issue. SFINAE checks for `tt::conforms_to` don't gain this advantage but also don't make errors worse: If the class indicates it conforms to the protocol but actually doesn't then the compiler hits the standard error once it tries to use a non-existing type alias, for example. Compiling the unit test should trigger more helpful error messages through `tt::assert_conforms_to`.
- By using protocols I hope devs define the compile-time interfaces more carefully and document them better.

Let me know your thoughts!

nilsdeppe commented on docs/DevGuide/Protocols.md:
```diff
@@ -71,41 +74,45 @@ placed in directory that also has the `greet` function.
+examples.
+
+Occasionally, you might be tempted to add template parameters to the
+protocol. In those situations, make the parameters part of your
```
I find that the first and second sentence in this paragraph seem to directly contradict one another. "You may want to do blah. But instead, you should do blah." Instead, I would've expected it to read "You may want to do blah. Instead, you should do X."

nilsdeppe commented on src/NumericalAlgorithms/DiscontinuousGalerkin/BoundarySchemes/FirstOrder/BoundaryData.hpp:
```diff
@@ -20,9 +20,8 @@ namespace FirstOrderScheme {
-  static_assert(tt::conforms_to_v<NumericalFluxType, protocols::NumericalFlux>,
-                "The 'NumericalFluxType' must conform to the "
-                "'dg::protocols::NumericalFlux'.");
+  static_assert(
```
Not a change request, but a point of discussion.

Here and in several other places for the numerical fluxes you check explicitly that they `assert_corform_to`. In contrast, the protocols documentation says
```
We are currently testing protocol conformance as part of our unit tests, so that
the `tt::conforms_to` metafunction only needs to check if a type inherits off
the protocol, but doesn't need to check the protocol's implementation. This is
both to keep compile times low and to have a SFINAE-friendly way of checking
protocol conformance. Full protocol conformance is tested with the
`tt::assert_conforms_to` metafunction.
```
which I interpreted as saying "You should only use `assert_conforms_to` in tests. How do you justify/remedy this apparent inconsistency?

nilsvu replied:
I didn't adjust that part of the documentation well enough. I think we can use `tt::assert_conforms_to` quite freely. Here's the re-formulated part of the documentation:

> The `tt::conforms_to` metafunction only checks if the class _indicates_ it
conforms to the protocol. Where SFINAE-friendliness is not necessary prefer the
`tt::assert_conforms_to` metafunction that triggers static asserts with
diagnostic messages to understand why the class does not conform to the
protocol.

Is that more clear?

nilsdeppe reviewed (commented):
I think this would address my concerns since, as you say, it shouldn't degrade compiler error quality, and if it does it's because somebody put bad error messages inside of an individual protocol, not a drawback of the entire design.

I have a couple minor comments but I'm happy for you to squash and rebase.

nilsvu committed:
Loosen requirements on numerical flux protocol

The package tags don't necessarily need to be DataBox tags.
At least the `package_extra_tags` are generally stored in a
`tuples::TaggedTuple`, so they can be any typed tag.
Closes #2204.

nilsvu committed:
Fix a few clang-tidy warnings
